lib/pwned_password_downloader.rb

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env ruby
+require 'fileutils'
+require 'net/http/persistent'
+require 'retries'
+require 'ruby-progressbar'
+require 'ruby-progressbar/outputs/null'
+
+class PwnedPasswordDownloader
+  attr_reader :destination,
+              :num_threads,
+              :keep_threshold,
+              :output_progress
+
+  alias_method :output_progress?, :output_progress
+
+  RANGE_API_ROOT = 'https://api.pwnedpasswords.com/range/'
+  SHA1_LENGTH = 40
+  HASH_PREFIX_LENGTH = 5
+  OCCURRENCE_OFFSET = SHA1_LENGTH - HASH_PREFIX_LENGTH + ':'.length
+
+  def initialize(
+    destination: 'tmp/pwned',
+    num_threads: 64,
+    keep_threshold: 30,
+    output_progress: true
+  )
+    @destination = destination
+    @num_threads = num_threads
+    @keep_threshold = keep_threshold
+    @output_progress = output_progress
+  end
+
+  def run!(start: '00000', finish: 'FFFFF')
+    (start.to_i(16)..finish.to_i(16)).each do |prefix_num|
+      queue << prefix_num.to_s(16).upcase.rjust(5, '0')
+    end
+
+    FileUtils.mkdir_p(destination)
+
+    progress_bar = ProgressBar.create(
+      title: 'Downloading...',
+      total: queue.size,
+      output: output_progress? ? $stdout : ProgressBar::Outputs::Null,
+      format: '[ %t ] %p%% %B %a (%E)',
+    )
+
+    [num_threads, queue.size].min.times.map do
+      Thread.new do |thread_id|
+        net_http = Net::HTTP::Persistent.new(name: "thread_id_#{thread_id}")
+
+        while (prefix = queue.pop)
+          if already_downloaded?(prefix)
+            progress_bar.increment
+            next
+          end
+
+          begin
+            write_one(
+              prefix:,
+              content: with_retries(
+                max_tries: HASH_PREFIX_LENGTH,
+                rescue: Socket::ResolutionError,
+              ) { download_one(prefix:, net_http:) },
+            )
+          rescue
+            queue << prefix
```

Comment:
Glancing at this quickly, with the loop, the max_tries, and then the "rescue queue << prefix" -- there might be the possibility of an infinite loop given a problem. Maybe this is more theoretical in the event of a network problem, but perhaps something to consider if this is run from some background job.

Reply:
Hm, yeah, @zachmargolis added this in d0863f18dbf00dbc46264dd40959e090553ca5a6, though I'm unsure about the `rescue` block adding back the prefix. It seems like it should bail out immediately for anything other than a socket error.

cc @zachmargolis if you have any thoughts / insights

Reply:
Yes as written, any non-socket error will cause it to bail out and re-add the prefix. I didn't want a thread to die and have the work not get picked up by another worker.

I'd also say, that since this script runs locally, and not that often, "best effort" is probably fine since it's restartable

So maybe we just ditch the rescue/retry entirely and have it print out a list of failed prefixes to be retried on another run

Reply:
I updated this in 4853620, including a few new test cases to cover successful retry and repeatedly failed retry.